### PR TITLE
feat: Deepspeed Trainer

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Ticket
https://hpe-aiatscale.atlassian.net/browse/MD-501

## Description
Add Deepspeed Trainer API
Refactor some pytorch training classes
Reinstate DCGANTrial as an example of DeepSpeed Trainer

## Test Plan
1. Try DCGAN trainer example. Instructions are in the `README` included with `tests/deepspeed/dcgan`. After it's done, continue the run for 100 more batches.
2. Run GPT NEOX example. After it's done, continue the run for 100 more batches.
3. Deepspeed unit tests pass.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py [hunk 1]
```diff
@@ -675,29 +1071,53 @@ def _save(self, path: pathlib.Path) -> None:
...
+                hparams = self.context.get_hparams()  # type: Optional[Dict[str, Any]]
+            except ValueError:
+                exp_conf = None
+                hparams = None
+
+            load_data = {
+                "trial_type": "DeepSpeedTrial",
+                "experiment_config": exp_conf,
+                "hparams": hparams,
+                "trial_cls_spec": f"{trial_cls.__module__}:{trial_cls.__qualname__}",
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_trial.py:1113]: so, this exists because we have `load_trial_from_checkpoint_path` in `pytorch/_load.py` which basically lets users load a checkpoint outside of a running trial, which had to be modified with PyTorch Trainer since users can instantiate their own trial classes now (with arbitrary arguments) and there's no easy way to load a checkpoint. 

i don't think we ever supported this sort of external checkpoint loading option for deepspeed, so i don't think it's worth doing now, but noting this as a potential future improvement.

but it means we don't actually need this in the checkpoint, or all the `is_trainer` stuff

File: examples/deepspeed/trainer/model_def.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
+import logging
+from typing import Any, Dict, Iterator, Optional, Tuple, Union, cast
+
+import data
+import deepspeed
+import torch
+import torch.nn as nn
+import torch.utils.data
+import torchvision
+from attrdict import AttrDict
```
Comment (azhou-determined) [inline examples/deepspeed/trainer/model_def.py:10]: import modules please

File: examples/deepspeed/trainer/model_def.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    overwrite_deepspeed_config,
+)
+
+REAL_LABEL = 1
+FAKE_LABEL = 0
+
+
+class DCGANTrial(DeepSpeedTrial):
+    def __init__(self, context: DeepSpeedTrialContext,
+                 hparams: AttrDict, data_config: AttrDict) -> None:
```
Comment (azhou-determined) [inline examples/deepspeed/trainer/model_def.py:26]: feel like `AttrDict` is kinda unnecessary here, just make it a plain dict?

File: examples/deepspeed/trainer/trainer.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    with det_ds.init() as train_context:
+        trial = DCGANTrial(train_context, hparams, data_config)
+        trainer = det_ds.Trainer(trial, train_context)
+        trainer.fit(max_length=max_length, latest_checkpoint=latest_checkpoint)
+
+if __name__ == "__main__":
+    args = parse_args()
+    # Configure logging
+    logging.basicConfig(level=logging.INFO, format=det.LOG_FORMAT)
+    main("mnist.yaml", local=args.local)
```
Comment (azhou-determined) [inline examples/deepspeed/trainer/trainer.py:55]: get rid of "mnist.yaml"

File: examples/deepspeed/trainer/mnist.yaml
Comment (azhou-determined) [inline examples/deepspeed/trainer/mnist.yaml:1]: fine to have this for testing, but since this is a legacy entrypoint, don't keep it in the examples

File: examples/deepspeed/trainer/trainer.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+from determined.pytorch import deepspeed as det_ds
+
+from model_def import DCGANTrial
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+
+    # Optional arguments for the launch helper
+    parser.add_argument("--local",
```
Comment (azhou-determined) [inline examples/deepspeed/trainer/trainer.py:18]: don't think this is needed. like, if you're running local, you'd just do `python3 trainer.py`. if you're on cluster, you'd call the same thing but from the config entrypoint.

in the code, use `local = det.get_cluster_info() is None`

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py [hunk 6]
```diff
@@ -646,27 +1031,38 @@ def _load(self, load_path: pathlib.Path) -> None:
                     "callback will be initialized from scratch"
                 )
 
-        # Load workload sequencer state.
-        wlsq_path = load_path.joinpath("workload_sequencer.pkl")
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_trial.py:650]: you need this. should test the case of loading a pre-deepspeed-trainer checkpoint into a post-deepspeed-trainer world

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py [hunk 7]
```diff
@@ -646,27 +1031,38 @@ def _load(self, load_path: pathlib.Path) -> None:
...
-            path.mkdir(parents=True, exist_ok=True)
-        _ = self.context.distributed.gather_local(None)  # sync
+        path.mkdir(parents=True, exist_ok=True)
 
         if self.is_chief:
             # We assume these stateful objects should be the same across slots and only have
             # the chief save them.
-            util.write_user_code(path, self.env.on_cluster)
-
-            if self.wlsq is not None:
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_trial.py:665]: >             # We assume these stateful objects should be the same across slots and only have
>             # the chief save them.

not a big deal, but previously we were saving the training state only on the chief (now TrialState)

File: examples/deepspeed/trainer/startup-hook.sh
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/startup-hook.sh:1]: Do we want this example to stay in? In which case I'll also remove the legacy Trial launch config (`mnist.yaml`).

File: harness/determined/pytorch/_util.py
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply harness/determined/pytorch/_util.py:1]: Well, user imports wouldn't change regardless if this is in `pytorch_trial.py` versus `_util.py`
What about renaming to `_training_utils.py`? I'd rather keep things that are shareable out of `pytorchtrial.py` rather than in it.

File: examples/deepspeed/trainer/startup-hook.sh
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/startup-hook.sh:1]: I see you concur in later comments lol

File: examples/deepspeed/trainer/model_def.py (hunk 3 above)
Comment (MikhailKardash) [reply examples/deepspeed/trainer/model_def.py]: Removed all traces of it from this example.

File: examples/deepspeed/trainer/trainer.py
Comment (azhou-determined) [inline examples/deepspeed/trainer/trainer.py:1]: in this file and a few others: your python files should have a blank newline at the end of each file. it's a PEP style, and it makes UNIX systems happier. also diffs can look weird without it.

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py [hunk 8]
```diff
@@ -65,15 +66,57 @@ class DeepSpeedTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
...
+        self,
+        core_context: det.core.Context,
+        trial_seed: Optional[int],
+        hparams: Optional[Dict],
+        slots_per_trial: int,
+        num_gpus: int,
+        exp_conf: Optional[Dict[str, Any]],
+        aggregation_frequency: int,
+        steps_completed: int,
+        debug_enabled: bool,
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_context.py:79]: this isn't getting used?

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py [hunk 9]
```diff
@@ -65,15 +66,57 @@ class DeepSpeedTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
...
+
+        self._trial_seed = trial_seed
+        self._steps_completed = steps_completed
+
         self._init_device()
 
         # Track which types we have issued warnings for in to_device().
         self._to_device_warned_types = set()  # type: Set[Type]
 
+        if aggregation_frequency > 1:
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_context.py:109]: there's no point is making this argument available if we're just going to say "do it in deepspeed instead"

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py (hunk 8 above)
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_deepspeed_context.py:79]: pls make a pass and look for params that aren't being used, these will be user-facing APIs, and if we decide to get rid of them later they'll be breaking changes.

File: examples/deepspeed/trainer/gan_model.py
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/gan_model.py:1]: This was due to bloat in our examples directory. It's just simple enough that reusing it for trainer made sense.

File: harness/determined/exec/harness.py [hunk 10]
```diff
@@ -167,6 +172,71 @@ def _run_pytorch_trial(
 
             trainer = pytorch.Trainer(trial_inst, train_context)
 
+            if "global_batch_size" in info.trial.hparams:
```
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply harness/determined/exec/harness.py:175]: looks like a combination of rebasing and changing `pytorch.Batch` -> `pytorch._Batch` and `pytorch.TrainUnit` -> `pytorch._TrainUnit`, then changing it back after. :|

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py [hunk 11]
```diff
@@ -166,12 +173,6 @@ def set_mpu(self, mpu: det_ds.ModelParallelUnit) -> None:
                 "Only one MPU can be passed to DeepSpeedTrialContext. "
                 "Please make sure wrap_mpu is only called once in the trial definition."
             )
-        if self.distributed.rank == 0:
```
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_deepspeed_context.py:169]: We got rid of `self._average_training_metrics`, so there's not need to do this check

File: examples/deepspeed/trainer/mnist_trainer.yaml
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/mnist_trainer.yaml:1]: In favor of calling it just `dcgan` since technically there are other datasets that are compatible with it (see `data.py`)

File: examples/deepspeed/trainer/requirements.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+mpi4py
```
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/requirements.txt]: I'll just move this to the `README.md`, since I don't force any `pip install` in local mode

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py (hunk 7 above)
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_deepspeed_trial.py:665]: DeepSpeedTrial previously called `_save()` on every rank, so each gpu had it's own state. The only thing done once was/is saving user code.

File: examples/deepspeed/trainer/README.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+## Files
+* **model_def.py**: The DCGANTrial definition.
+* **gan_model.py**: Network definitions for generator and discriminator.
+* **data.py**: Dataset loading/downloading code.
+
+### Configuration Files
+* **ds_config.json**: The DeepSpeed config file.
+* **mnist.yaml**: Determined config to train the model on mnist.
+* **mnist_grad_accum.yaml**: Same as `mnist.yaml`, but with gradient accumulation turned on.
+* **cifar10_zero2.yaml**: Determined config to train the model on `cifar10` with ZeRO stage 2 and float16 mxied precision enabled.
```
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply examples/deepspeed/trainer/README.md]: I added a running locally section

File: harness/determined/pytorch/_util.py
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/pytorch/_util.py:1]: > Well, user imports wouldn't change regardless if this is in pytorch_trial.py versus _util.py

yeah, that's what i'm saying, user imports are the same anyway so it's a purely internal concern that we move it. i was hoping that deepspeed trial could subclass pytorchtrial at some point, so i guess that's how i was thinking about it.

i'm fine either way on this. what about `_trainer_utils.py`? feel like that's slightly better than `_util`.

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py (hunk 7 above)
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_deepspeed_trial.py:665]: but the saving of the training state is also gated by the chief worker?

```
        if self.is_chief:
            # We assume these stateful objects should be the same across slots and only have
            # the chief save them.
            util.write_user_code(path, self.env.on_cluster)

            if self.wlsq is not None:
                with path.joinpath("workload_sequencer.pkl").open("wb") as f:
                    pickle.dump(self.wlsq.get_state(), f)

```

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py (hunk 9 above)
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_deepspeed_context.py]: you resolved this but nothing has changed? 👀

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py [hunk 14]
```diff
@@ -65,15 +66,57 @@ class DeepSpeedTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
...
+        if aggregation_frequency > 1:
+            raise det.errors.InvalidExperimentException(
+                "Gradient aggregation is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+        self._aggregation_frequency = aggregation_frequency
+
+        self._fp16_compression_default = False
+        self._average_aggregated_gradients_default = True
+        self._is_pre_trainer = False
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_context.py]: shouldn't be used, this is an artifact of the external checkpoint loading thing i was talking about earlier.

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py [hunk 15]
```diff
@@ -272,6 +272,15 @@ def _init_device(self) -> None:
             self.device = torch.device("cuda", 0)
         assert self.device is not None, "Error setting torch device."
 
+    def _set_default_gradient_compression(self, gradient_compression: bool) -> None:
+        self._fp16_compression_default = gradient_compression
+
+    def _set_default_average_aggregated_gradients(self, average_aggregated_gradients: bool) -> None:
+        self._average_aggregated_gradients_default = average_aggregated_gradients
+
+    def _set_is_pre_trainer(self) -> None:
```
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_context.py]: this shouldn't be needed if we're not doing the external checkpoint loading thing

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py (hunk 6 above)
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_deepspeed_trial.py:650]: you resolved this but i still don't see the legacy workload seq checkpoint being loaded?

File: harness/determined/pytorch/deepspeed/_trainer.py [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+        )
+
+        trial_controller.run()
+
+
+def _initialize_distributed_backend() -> Optional[core.DistributedContext]:
+    info = det.get_cluster_info()
+
+    if torch.cuda.is_available():
+        deepspeed.init_distributed()
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_trainer.py:186]: wait, but i think the whole point of that manual flag is to _not_ do this if DET_MANUAL_INIT_DISTRIBUTED is false.

[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_trainer.py:186]: this whole block would be easier read as:

```
if distributed_backend.use_deepspeed():
    manual_dist_init = os.environ.get("DET_MANUAL_INIT_DISTRIBUTED")
    if not manual_dist_init:
        deepspeed.init_distributed(auto_mpi_discovery=False)
    return core.DistributedContext.from_deepspeed()
```

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py (hunk 9 above)
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_deepspeed_context.py]: nvm i was looking at an old commit lol

File: harness/determined/pytorch/deepspeed/_trainer.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_trainer.py:186]: I believe this would break my example, which can be run on gpu without explicitly calling the launcher, via `python trainer.py`

[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/pytorch/deepspeed/_trainer.py:186]: this should only get called if local training != true tho

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py
Comment (azhou-determined) [inline harness/determined/pytorch/deepspeed/_deepspeed_context.py:1]: feel like the `num_micro_batches_per_slot` and `train_micro_batch_size_per_gpu` should be `get_*` following convention on the rest of the methods here.

**Comment by azhou-determined** [comment]
the gpt_neox example still uses the legacy harness trial. we should either get rid of the example if it's not useful, or port it to trainer

**Review by dannysauer — APPROVED** [review]
Infra part is fine

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: harness/determined/pytorch/deepspeed/_deepspeed_trial.py (hunk 7 above)
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_deepspeed_trial.py:665]: added state to the top `self.is_chief` portion of this method

File: harness/determined/pytorch/deepspeed/_trainer.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_trainer.py]: updated

**Review by jesse-amano-hpe — APPROVED** [review]

File: harness/determined/pytorch/deepspeed/_deepspeed_context.py
Comment (MikhailKardash) [reply harness/determined/pytorch/deepspeed/_deepspeed_context.py:1]: replaced `@property` with `get_`
